# GAARD GWAS results summary

Each plot shows the 1000 most significant SNPs in the GWAS for that sample set. At the top, the P value for each SNP is shown, colour-coded by chromosome. Underneath is a heatmap showing the pairwise correlation between all 1000 SNPs, helping to see groups of SNPs that are in linkage disequilibrium. Each point in the *P*-value plot is lined up with its column in the heatmap. Beneath the column, lines heatmap columns to a map of the genome to show the genomic position of each SNP. Black horizontal bars between the heatmap and the linking lines indicate genomic regions containing a high density of these top 1000 SNPs (100 kb windows that contained at least 10 SNPs).

&nbsp;
&nbsp;

___

[Moshi\_*arabiensis*\_Delta](#moshi_arabiensis_delta)  
[Muleba\_*arabiensis*\_Delta](#muleba_arabiensis_delta)  
[Moshi\_*arabiensis*\_PM](#moshi_arabiensis_pm)  

___

## Moshi\_*arabiensis*\_Delta

![Moshi_arabiensis_Delta_top_snps](/home/elucas/data/ML/GAARD_SNP/summary_figures/Moshi_arabiensis_Delta_top_snps_summary_plot.png)

&nbsp;

___

## Muleba\_*arabiensis*\_Delta

![Muleba_arabiensis_Delta_top_snps](/home/elucas/data/ML/GAARD_SNP/summary_figures/Muleba_arabiensis_Delta_top_snps_summary_plot.png)


&nbsp;

___

## Moshi\_*arabiensis*\_PM

![Moshi_arabiensis_PM_top_snps](/home/elucas/data/ML/GAARD_SNP/summary_figures/Moshi_arabiensis_PM_top_snps_summary_plot.png)

&nbsp;

___
